webrtc/README.md

```diff
@@ -0,0 +1,439 @@
+framing. Messages on a `RTCDataChannel` are embedded into the Protobuf message
+below and sent on the `RTCDataChannel` prefixed with the message length in
+bytes, encoded as an unsigned variable length integer as defined by the
+[multiformats unsigned-varint spec][uvarint-spec].
+
+It is an adaptation from the [QUIC RFC]. When in doubt on the semantics of
+these messages, consult the [QUIC RFC].
+
+``` proto
+syntax = "proto2";
+
+package webrtc.pb;
+
+message Message {
+  enum Flag {
+    // The sender will no longer send messages on the stream.
+    FIN = 0;
+    // The sender will no longer read messages on the stream. Incoming data is
+    // being discarded on receipt.
+    STOP_SENDING = 1;
+    // The sender abruptly terminates the sending part of the stream. The
+    // receiver can discard any data that it already received on that stream.
+    RESET_STREAM = 2;
+  }
+
+  optional Flag flag=1;
+
+  optional bytes message = 2;
+}
+```
+
+Note that in contrast to QUIC (see [QUIC RFC - 3.5 Solicited State
+Transitions](https://www.rfc-editor.org/rfc/rfc9000.html#section-3.5)) a libp2p
+WebRTC endpoint receiving a `STOP_SENDING` frame SHOULD NOT send a
+`RESET_STREAM` frame in reply. The `STOP_SENDING` frame is used for accurate
+accounting of the number of bytes sent for connection-level flow control in
+QUIC. The libp2p WebRTC message framing is not concerned with flow-control and
+thus does not need the `RESET_STREAM` frame to be send in reply to a
+`STOP_SENDING` frame.
+
+Encoded messages including their length prefix MUST NOT exceed 16kiB to support
+all major browsers. See ["Understanding message size
+limits"](https://developer.mozilla.org/en-US/docs/Web/API/WebRTC_API/Using_data_channels#understanding_message_size_limits).
+Implementations MAY choose to send smaller messages, e.g. to reduce delays
+sending _flagged_ messages.
+
+### Ordering
+
+Implementations MAY expose an unordered byte stream abstraction to the user by
+overriding the default value of `ordered` `true` to `false` when creating a new
+data channel via
+[`RTCPeerConnection.createDataChannel`](https://www.w3.org/TR/webrtc/#dom-peerconnection-createdatachannel).
+
+### Open Questions
+
+- There are two ways to open a `RTCDataChannel`, either via `negotiated: true`
+  and thus out-of-band channel ID negotiation or `negotiated: false` and thus
+  in-band channel ID negotiation.
+
+  Is the assumption correct that in-band stream ID negotiation requires an
+  additional round trip? If so how about using out-of-band channel id
+  negotiation only (thus no additional round trip) based on the DTLS role?
+
+  > When one side wants to open a channel using out-of-band negotiation, it
+  > picks a stream. Unless otherwise defined or negotiated, the streams are
+  > picked based on the DTLS role (the client picks even stream identifiers, and
+  > the server picks odd stream identifiers).
+
+  https://www.rfc-editor.org/rfc/rfc8831#name-opening-a-data-channel
```

Comment:
@ckousik @lightsofapollo or anyone else: Do you know more of the inner working on data channel ID negotiation in the major browsers and SCTP in general? If not, I can do some more digging.

_Sparked by @marten-seemann comment in https://github.com/libp2p/specs/pull/412#issuecomment-1255472209._

Reply:
The peer would still have to signal to the remote that a new datachannel is being used, so that the remote can instantiate the datachannel.

Reply:
@melekes what if both peers choose the same channel ID? That would be a race condition, right? How can SCTP prevent that race condition without a full round trip?

Reply:
This is now documented with 3315b5c. Let me know what you think.

Reply:
> That would be a race condition, right?

I'll have to read the source code, but from my understanding the peer who already has channel with given ID will just reject the instruction to create it.

Reply:
> > Is the assumption correct that in-band stream ID negotiation requires an additional round trip?
> 
> I don't think it's a "round trip", more like _A_ sends a message to _B_ telling it to open a channel with a certain ID.

https://datatracker.ietf.org/doc/html/draft-ietf-rtcweb-data-protocol-09#section-6 describes the odd/even rule for peers to pick a streamID. While a peer can send messages before the round trip is completed, it will wait for an ack message before changing the data channel's readyState to `open`.

Reply:
@ckousik thanks for the correction. Honestly, I'm still confused about this item. The naive implementation, where a client just creates a data channel with an ID generated according to rules ^ and `negotiated: true`, doesn't work (at least not in Rust / webrtc-rs). 

> If a DATA_CHANNEL_OPEN message is received on an unused Stream

if negotiated is true, no DATA_CHANNEL_OPEN msg is sent (I think it's true for both webrtc-rs and Pion).

how should we inform the remote about the new channel?

Reply:
This is very unfortunate. @ckousik can you confirm this for pion? @John-LittleBearLabs can you confirm this on the browser side?

Worst case we have to do `negotiated: false`, hope for the user-agent to re-use channel ids or be bound to 2^16 channels, and embrace the fact that the channel listener (instead of the initiator) is able to send data first (after having accepted the inbound channel request).

Reply:
Yes, but `negotiated: true` is only meant for cases where we know in advance that a datachannel is going to exist and with a given ID. In our use case, we can only do `negotiated: true` for the handshake channel, and every subsequent channel has to be `negotiated: false`.

Reply:
faf641d updates the specification accordingly, i.e. requires `negotiated: false`. Note that this does not have to imply an additional round trip on each new `RTCDataChannel`. See commit.

Your input was very helpful here @melekes and @ckousik!

Reply:
Resolving here. Very much appreciate all the input.